core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Manager.java

```diff
@@ -0,0 +1,239 @@
+    super(name + "-token-refresh");
+  }
+
+  @Override
+  public OAuth2Util.AuthSession initSession(RESTClient initClient, Map<String, String> properties) {
+    warnIfDeprecatedTokenEndpointUsed(properties);
+    AuthConfig config = AuthConfig.fromProperties(properties);
+    Map<String, String> headers = OAuth2Util.authHeaders(config.token());
+    OAuth2Util.AuthSession session = new OAuth2Util.AuthSession(headers, config);
+    if (config.credential() != null && !config.credential().isEmpty()) {
+      // keep track of the start time for token refresh
+      this.startTimeMillis = System.currentTimeMillis();
+      this.authResponse =
+          OAuth2Util.fetchToken(
+              initClient,
+              headers,
+              config.credential(),
+              config.scope(),
+              config.oauth2ServerUri(),
+              config.optionalOAuthParams());
+      return OAuth2Util.AuthSession.fromTokenResponse(
+          initClient, null, authResponse, startTimeMillis, session);
+    } else if (config.token() != null) {
+      return OAuth2Util.AuthSession.fromAccessToken(
+          initClient, null, config.token(), null, session);
+    }
+    return session;
+  }
+
+  @Override
+  public OAuth2Util.AuthSession catalogSession(
+      RESTClient sharedClient, Map<String, String> properties) {
+    this.client = sharedClient;
+    this.sessionCache = new AuthSessionCache(sessionTimeout(properties));
+    AuthConfig config = AuthConfig.fromProperties(properties);
+    Map<String, String> headers = OAuth2Util.authHeaders(config.token());
+    OAuth2Util.AuthSession session = new OAuth2Util.AuthSession(headers, config);
+    keepRefreshed(config.keepRefreshed());
+    // authResponse comes from the init phase
+    if (authResponse != null) {
+      return OAuth2Util.AuthSession.fromTokenResponse(
+          client, refreshExecutor(), authResponse, startTimeMillis, session);
+    } else if (config.token() != null) {
+      return OAuth2Util.AuthSession.fromAccessToken(
+          client, refreshExecutor(), config.token(), config.expiresAtMillis(), session);
+    }
+    return session;
+  }
+
+  @Override
+  public OAuth2Util.AuthSession contextualSession(
+      SessionCatalog.SessionContext context, AuthSession parent) {
+    return maybeCreateChildSession(
+        context.credentials(),
+        context.properties(),
+        ignored -> context.sessionId(),
+        (OAuth2Util.AuthSession) parent);
+  }
+
+  @Override
+  public OAuth2Util.AuthSession tableSession(
+      TableIdentifier table, Map<String, String> properties, AuthSession parent) {
+    return maybeCreateChildSession(
+        Maps.filterKeys(properties, TABLE_SESSION_ALLOW_LIST::contains),
+        properties,
+        properties::get,
+        (OAuth2Util.AuthSession) parent);
+  }
+
+  @Override
+  public void close() {
+    try {
+      super.close();
+    } finally {
+      AuthSessionCache cache = sessionCache;
+      sessionCache = null;
```

Comment:
nit: should be `this.sessionCache` to clearly indicate that an instance variable is assigned here (and to align with other places in the codebase